group_theory/coset.lean

```diff
@@ -148,44 +158,69 @@ local attribute [instance] left_rel
 
 noncomputable def group_equiv_left_cosets_times_subgroup (hs : is_subgroup s) :
   α ≃ (left_cosets s × s) :=
-calc α ≃ Σ L : left_cosets s, {x // ⟦x⟧ = L} :
-  equiv.equiv_fib quotient.mk
-    ... ≃ Σ L : left_cosets s, left_coset (quotient.out L) s :
-  equiv.sigma_congr_right (λ L, by rw ← left_cosets.eq_class_eq_left_coset; simp)
+calc α ≃ Σ L : left_cosets s, {x : α // (x : left_cosets s)= L} :
+  equiv.equiv_fib left_cosets.mk
+    ... ≃ Σ L : left_cosets s, left_coset (quotient.out' L) s :
+  equiv.sigma_congr_right (λ L, 
+    begin rw ← left_cosets.eq_class_eq_left_coset, 
+      show {x // quotient.mk' x = L} ≃ {x : α // quotient.mk' x = quotient.mk' _},
+      simp [-quotient.eq']
+    end)
     ... ≃ Σ L : left_cosets s, s :
   equiv.sigma_congr_right (λ L, left_coset_equiv_subgroup _)
     ... ≃ (left_cosets s × s) :
   equiv.sigma_equiv_prod _ _
 
 end is_subgroup
 
-section quotient_group
-
-local attribute [instance] left_rel normal_subgroup.to_is_subgroup
+namespace left_cosets
 
 instance [group α] (s : set α) [normal_subgroup s] : group (left_cosets s) :=
```

Comment:
I'm wondering if you could shrink this instance using `refine_struct` (like in https://github.com/leanprover/mathlib/blob/master/algebra/pi_instances.lean)

Reply:
I gave it try, I might have a substitute proof for you in a few moments.

Reply:
Have a look at https://github.com/dorhinj/mathlib/pull/9 and see if you find that it improves your PR.

Reply:
I'm not sure it's much of an improvement, it makes the proof harder to understand as far as I can see. I'll let the people who have to maintain it decide if it's an improvement.

Reply:
For only one proof it might not be an improvement, you're right but if you have a pattern of writing algebraic structures on quotient types, repeating the same proof over and over might not be too enlightening.

Reply:
I'm sure using `refine_struct` and a more or less specialize tactic is a good idea. I don't understand why Simon took the definition of `mul` and `inv` out of the initial structure, but the important part is https://github.com/cipher1024/mathlib/blob/7f14aa5c85f04645ba41fa4740e549a4efa2ca47/group_theory/coset.lean#L222-L225 which could be made into a `derive_quotient_field` tactic and reused in a lot of places.

Reply:
I didn't mean those functions should be automated. Every automation is good of course, but the first targets here were clearly those repeated `(λ ..., congr_arg mk (some_field ...))`.